contracts/tgrade-valset/src/contract.rs

```diff
@@ -537,8 +551,22 @@ fn end_block(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
         JAIL.remove(deps.storage, addr)
     }
 
-    let old_validators = VALIDATORS.load(deps.storage)?;
-    VALIDATORS.save(deps.storage, &validators)?;
+    // First iteration - load all validators, second - remove all, third - add all new
+    // Not very efficient...
+    let old_validators = VALIDATORS
+        .range(deps.storage, None, None, Order::Descending)
+        .map(|validator_info| {
+            let (_, info) = validator_info?;
+            Ok(info)
+        })
+        .collect::<StdResult<Vec<ValidatorInfo>>>()?;
+    for validator in &old_validators {
+        VALIDATORS.remove(deps.storage, &validator.operator);
+    }
+    for validator in &validators {
+        VALIDATORS.save(deps.storage, &validator.operator, validator)?;
+    }
```

Comment:
@maurolacy worked on something similar with HashMap in some other PR

Reply:
I think BTree map was important if we want deterministic sorting order.

Can you link the other contract that did this?

Reply:
There are two things here:

1. Pagination. This is (relatively) minor, and can be solved in different ways.
2. `Item` value size limitations. This is something that came up during discussion, and it's important, because we're storing `ValidatorMetadata`, which has a lot of `String` fields, inside `ValidatorInfo`. These fields are also in control of the operator, so, this can be clearly abused, and can lead to performance issues, or even, denial of service.

We'll tackle 2. in #42.

Regarding 1., and without changing the underlying structure, a simple approach would be to populate a `BtreeSet` with the data, and then range over it using [BTreeSet::range](https://doc.rust-lang.org/stable/std/collections/struct.BTreeSet.html#method.range). and paginate using `take` over the iterator as usual.

A similar / equivalent way would be to sort the vec, and then bisect, but I think the btree approach is simpler / better.

Reply:
Another option could be to turn `VALIDATORS` into an `Item<BTreeSet<ValidatorInfo>>` structure.

Need to research this a bit, but at first glance it looks overkill. We will also have to be very careful with the size requirements of `BTreeSet` itself. And, reducing the `ValidatorInfo` footprint as well, like mentioned in #42.

Reply:
Finally, implementing the `Map` approach for `VALIDATORS` doesn't look like a bad idea either. It will solve both the pagination, and the size / limits issue. At the expense of some required code changes / adaptations.

In that case, the way to make the code above efficient would be to process only the updates and removals over the **differences**, i.a. as part of (or after) `calculate_diff`.